ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs

```diff
@@ -0,0 +1,104 @@
+{-# LANGUAGE TemplateHaskell           #-}
+
+module Development.IDE.Plugin.CodeAction.Args (
+  module Development.IDE.Plugin.CodeAction.Args,
+) where
+
+import           Control.Lens                                 (alaf)
+import           Data.Bifunctor                               (second)
+import           Data.Monoid                                  (Ap (..))
+import qualified Data.Text                                    as T
+import           Development.IDE
+import           Development.IDE.GHC.Compat
+import           Development.IDE.Plugin.CodeAction.Args.TH
+import           Development.IDE.Plugin.CodeAction.ExactPrint
+import           Development.IDE.Plugin.TypeLenses            (GlobalBindingTypeSigsResult)
+import           Development.IDE.Spans.LocalBindings          (Bindings)
+import           Development.IDE.Types.Exports                (ExportsMap)
+import           Development.IDE.Types.Options                (IdeOptions)
+import           Language.LSP.Types                           (TextEdit,
+                                                               type (|?) (..))
+import           Retrie                                       (Annotated (astA))
+import           Retrie.ExactPrint                            (annsA)
+
+data CodeActionArgs = CodeActionArgs
+  { caaExportsMap   :: ExportsMap
+  , caaIdeOptions   :: IdeOptions
+  , caaParsedModule :: Maybe ParsedModule
+  , caaContents     :: Maybe T.Text
+  , caaDf           :: Maybe DynFlags
+  , caaAnnSource    :: Maybe (Annotated ParsedSource)
+  , caaTmr          :: Maybe TcModuleResult
+  , caaHar          :: Maybe HieAstResult
+  , caaBindings     :: Maybe Bindings
+  , caaGblSigs      :: Maybe GlobalBindingTypeSigsResult
+  , caaDiagnostics  :: Diagnostic
+  }
+
+rewrite ::
+  Maybe DynFlags ->
+  Maybe (Annotated ParsedSource) ->
+  [(T.Text, [Rewrite])] ->
+  [(T.Text, [TextEdit])]
+rewrite (Just df) (Just ps) r
+  | Right edit <-
+      (traverse . traverse)
+        (alaf Ap foldMap (rewriteToEdit df (annsA ps)))
+        r =
+    edit
+rewrite _ _ _ = []
+
+-- we need this intermediate existential type to encapsulate functions producing code actions into a list
+data SomeAction = forall a. ToCodeAction a => SomeAction a
+
+wrap :: ToCodeAction a => a -> SomeAction
+wrap = SomeAction
+
+unwrap :: CodeActionArgs -> SomeAction -> [(T.Text, [TextEdit])]
+unwrap caa (SomeAction x) = toCodeAction caa x
+
+class ToCodeAction a where
+  toCodeAction :: CodeActionArgs -> a -> [(T.Text, [TextEdit])]
+
+instance ToCodeAction [(T.Text, [TextEdit])] where
+  toCodeAction _ = id
+
+instance ToCodeAction [(T.Text, [Rewrite])] where
+  toCodeAction CodeActionArgs{..} = rewrite caaDf caaAnnSource
+
+instance ToCodeAction r => ToCodeAction (ParsedSource -> r) where
+  toCodeAction caa@CodeActionArgs{caaAnnSource = Just ps} f = toCodeAction caa $ f $ astA ps
+  toCodeAction _ _ = []
+
+instance ToCodeAction [(T.Text, [TextEdit |? Rewrite])] where
```

Comment:
isn't `|?` just [`Sum`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Functor-Sum.html#t:Sum) which is available in the Prelude?

Reply:
I think it's more like `Either` with special JSON treatment, as specified at https://github.com/wz1000/haskell-lsp/blob/master/lsp-types/src/Language/LSP/Types/Common.hs

Reply:
Is the special JSON treatment relevant here? If not, why not use `Either`?

Reply:
No. I thought commonly `Left` was used to express the error, so I chose the in scope `|?`. But now I think `Either TextEdit Rewrite` makes sense: we can't express this modification in `Rewrite`, thus we have to use `TextEdit` as the fallback.